### PR TITLE
Update an example on transformer_in_transformer

### PR DESCRIPTION
**Pull request by czy00000**
sove the problem of this PR #687

### CODE REVIEW

File: examples/vision/transformer_in_transformer.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    # pixel extraction
+    x = tf.image.extract_patches(
+        images=x,
+        sizes=(1, new_patch_size[0], new_patch_size[1], 1),
+        strides=(1, new_patch_size[0], new_patch_size[1], 1),
+        rates=(1, 1, 1, 1),
+        padding="VALID",
+    )
+    x = tf.reshape(x, shape=(-1, new_patch_size[0] * new_patch_size[1], in_dim))
+    x = PatchEncoder(new_patch_size[0] * new_patch_size[1], in_dim)(x)
```
Comment (czy00000) [inline examples/vision/transformer_in_transformer.py:121]: # disgard tf.Variable

**Comment by innat** [comment]
@czy00000 
It would be very useful to add the following two sections to the transformer-based code example. 

- Add TPU support 
- Class-Activation Maps (compare to complementary CNN approach).

**Comment by czy00000** [comment]
@fchollet Can someone help to review this PR?

**Comment by czy00000** [comment]
@innat Thanks for your advice, the focus of this example is only how to effectively implement this structure. TPU support and class activation map implementation can be found in other examples.

**Comment by mattdangerw** [comment]
Thanks for the PR! I will take a look later today.

File: examples/vision/transformer_in_transformer.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    residual_in_2 = pixel_embedding
+    pixel_embedding = layers.LayerNormalization(epsilon=1e-5)(pixel_embedding)
+    pixel_embedding = mlp(
+        pixel_embedding, in_embedding_dim * mlp_ratio, in_embedding_dim
+    )
+    pixel_embedding = layers.add([pixel_embedding, residual_in_2])
+
+    # outer transformer block
+    _, num_patches, channel = patch_embedding.shape
+    # fusion local and global information
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:171]: fuse local and global information

File: examples/vision/transformer_in_transformer.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    fusion_embedding = tf.reshape(
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:172]: fusion_embedding -> fused_embedding

File: examples/vision/transformer_in_transformer.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+The TNT model consists of multiple TNT blocks.
+In the TNT block, there are two transformer blocks where
+the outer transformer block models the global relation among patch embeddings,
+and the inner one extracts local structure information of pixel embeddings.
+The local information is added on the patch
+embedding by linearly projecting the pixel embeddings into the space of patch embedding.
+Patch-level and pixel-level position embeddings are introduced in order to
+retain spatial information.In orginal paper, the authors use the class token for
+classification.
+We use the `layers.GlobalAvgPool1D` to fusion patch information.
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:204]: to fusion patch information -> to fuse patch information

File: examples/vision/transformer_in_transformer.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+"""
+
+
+def get_model(
+    image_size=96,
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:209]: You are repeating a number of constants here that were defaulted in your helper functions. Either place is fine, but don't duplicate these numbers. It makes it hard what to tell what was used (e.g. outer_block_num_heads is 4 here, and default to out_num_heads is 8 in transformer_in_transformer_block).

File: examples/vision/transformer_in_transformer.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+print(f"Test top 5 accuracy: {round(top_5_accuracy * 100, 2)}%")
+
+"""
+After 50 epochs, the TNT model achieves around 48% accuracy and
+78% top-5 accuracy on the test data. It only has 0.8M parameters.
+From the above loss curve, we can find that the model gradually converges,
+but it never achieves state of the art performance. We can apply some data augmentation
+to
+obtain better performance, like RandAugment, MixUp etc. We also can adjust the depth
+of model, learning rate or increase the size of embedding.Compared to the conventional
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:323]: typo, space after period

File: examples/vision/transformer_in_transformer.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+vision transformers [(ViT)](https://arxiv.org/abs/2010.11929) which corrupts the local
+structure
+of the patch, the TNT can better preserve and model the local information
+for visual recognition.
+"""
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:328]: If you know of any good links for further reading to go deeper, feel free to include them at the end of the example.

File: examples/vision/transformer_in_transformer.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+loss, accuracy, top_5_accuracy = model.evaluate(x_test, y_test)
+print(f"Test loss: {round(loss, 2)}")
+print(f"Test accuracy: {round(accuracy * 100, 2)}%")
+print(f"Test top 5 accuracy: {round(top_5_accuracy * 100, 2)}%")
+
+"""
+After 50 epochs, the TNT model achieves around 48% accuracy and
+78% top-5 accuracy on the test data. It only has 0.8M parameters.
+From the above loss curve, we can find that the model gradually converges,
+but it never achieves state of the art performance. We can apply some data augmentation
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:320]: We can apply some data augmentation -> We could apply further data augmentation

File: examples/vision/transformer_in_transformer.py [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+to
+obtain better performance, like RandAugment, MixUp etc. We also can adjust the depth
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:322]: Link to sources on these two approaches, otherwise this will confuse people as to what RandAugment, MixUp are

File: examples/vision/transformer_in_transformer.py [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+validation_split = 0.2
+batch_size = 128
+image_size = 96  # resize images to this size
+num_epochs = 50
+
+"""
+## Use data augmentation
+"""
+
+data_augmentation = keras.Sequential(
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:71]: It is a little weird you are using a sequential model here when elsewhere you are using

```
some_function(input):
  return output
```

Applied to function model inputs. Why not just do the same here, and define a function which applies these layers to some functional input?

File: examples/vision/transformer_in_transformer.py [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+    inputs = layers.Input(shape=input_shape)
+    # Image augment
+    x = data_augmentation(inputs)
+    # extract pixel embedding
+    pixel_embedding, num_patches, new_patch_size = pixel_embed(
+        x, image_size, patch_size, inner_block_embedding_dim, first_stride
+    )
+    num_pixel = new_patch_size[0] * new_patch_size[1]
+    # extract patch embedding
+    patch_embedding = tf.reshape(
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:230]: I think this would read clearer if you refactored this part which computes the patch_embedding into it's own patch_embed function (similar to pixel_embed).

File: examples/vision/transformer_in_transformer.py [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+        encoded = self.projection(patch) + self.position_embedding(positions)
+        return encoded
+
+
+def pixel_embed(x, image_size=96, patch_size=8, in_dim=48, stride=4):
+    _, channel, height, width = x.shape
+    image_size = tuple(repeat(image_size, 2))
+    patch_size = tuple(repeat(patch_size, 2))
+    num_patches = (image_size[0] // patch_size[0]) * (image_size[1] // patch_size[1])
+    new_patch_size = [math.ceil(ps / stride) for ps in patch_size]
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:110]: new_patch_size is not a very descriptive name. ideas for a better one?

File: examples/vision/transformer_in_transformer.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+    outputs = layers.Dense(num_classes, activation="softmax")(x)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    return model
+
+
+"""
+## Train on CIFAR-100
+"""
+
+model = get_model()
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:265]: displaying the output of model.summary() somewhere might be helpful

File: examples/vision/transformer_in_transformer.py [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+"""
+## Implement the TNT model
+The TNT model consists of multiple TNT blocks.
+In the TNT block, there are two transformer blocks where
+the outer transformer block models the global relation among patch embeddings,
+and the inner one extracts local structure information of pixel embeddings.
+The local information is added on the patch
+embedding by linearly projecting the pixel embeddings into the space of patch embedding.
+Patch-level and pixel-level position embeddings are introduced in order to
+retain spatial information.In orginal paper, the authors use the class token for
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:202]: typo: space after period

File: examples/vision/transformer_in_transformer.py [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+Title: Image classification with TNT(Transformer in Transformer)
+Author: [ZhiYong Chang](https://github.com/czy00000)
+Date created: 2021/10/25
+Last modified: 2021/11/06
+Description: Implementing the Transformer in Transformer (TNT) model for image classification.
+"""
+
+"""
+## Introduction
+This example implements the [TNT](https://arxiv.org/abs/2103.00112)
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:11]: This looks like it is not precisely the architecture hyper-parameters from the paper correct? (depth of 5 blocks vs 12?)

If that's right, might be helpful to introduce that somewhere in this introduction. "To keep training time reasonable will train and test a smaller model than is in the paper (XX params vs XX params)." Something like that.

File: examples/vision/transformer_in_transformer.py [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+model for image classification, and demonstrates it on the CIFAR-100 dataset.
+TNT is a novel model for modeling both patch-level and pixel-level
+representation. In each TNT block, an ***outer*** transformer block is utilized to
+process
+patch embeddings, and an ***inner***
+transformer block extracts local features from pixel embeddings. The pixel-level
+feature is projected to the space of patch embedding by a linear transformation layer
+and then added into the patch.
+This example requires TensorFlow 2.5 or higher, as well as
+[TensorFlow Addons](https://www.tensorflow.org/addons/overview) package,
```
Comment (mattdangerw) [inline examples/vision/transformer_in_transformer.py:21]: mention that this is for the AdamW optimizer

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks very much for the new example! This overall reads very clearly and is a nice example. Left some comments.

**Comment by czy00000** [comment]
@mattdangerw Thank you for taking the time to review! I will update this example.

**Comment by czy00000** [comment]
@mattdangerw I have updated this file according to your advice.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks! Few more comments. Do you have a colab version of this to try running?

**Comment by czy00000** [comment]
@mattdangerw Thanks for your advice! I have added the jupyter version and generated files.

**Comment by mattdangerw** [comment]
@czy00000 Thanks for making the changes. I looked this over as a colab, and I think this could use some more text explaining what is happening in the code blocks. Right now you introduce the problem, but then give a lot of code without explanation until the "Implement the TNT model" block. In particular the sections below could do with an introductory short paragraph explaining what the code will be doing:

 - Implement the pixel embedding and patch embedding layer
 - Implement the MLP block
 - Implement the TNT block

Other than that, the other place I am still unsure is performance. @fchollet I know you had worries about the previous incarnation https://github.com/keras-team/keras-io/pull/687. Performance went from `35% accuracy and
66% top-5 accuracy` -> `42% accuracy and
73% top-5 accuracy`, is this still too low?

**Comment by fchollet** [comment]
> 42% accuracy and 73% top-5 accuracy

These are decent results for CIFAR-100 for a tutorial. Thanks for checking.
